### ISSUE TITLE
Import creates two image instances when xmp filename ends "_01.cr2.xmp"

### ISSUE THREAD

**Issue by elstoc** [darktable-org/darktable#7957]
Import a folder that contains a previously edited image and an xmp file. If the xmp file has a filename "<img_filename>_01.cr2.xmp" darktable creates two duplicate instances of the image and generates a second xmp file (named "<img_filename>.cr2.xmp"). Same thing happens if I rename it to "<img_filename>_02.cr2.xmp".

If the xmp is named "<img_filename>.cr2.xmp" only a single instance of the image is imported.

This seems to be a regression of an issue I originally raised and resolved: #4385

3.5.0+800~g3366c75b9
ArchLinux

**Comment by elstoc** [comment]
This looks to me like `dt_image_find_duplicates` is somehow finding an extra non-existent file but I really don't know how it works.  If I look at the `version` and `xmpfilename` variables in the `file_iter` loop in `_image_read_duplicates` though, it returns:
```
version 0 
file /home/chris/Photos/00 All Photos/000 Darktable Master Testing/2017-04-21-11.56.04.cr2.xmp 
version 3 
file /home/chris/Photos/00 All Photos/000 Darktable Master Testing/2017-04-21-11.56.04_03.cr2.xmp 
```
even though the v0 xmp does not exist prior to import

**Comment by elstoc** [comment]
@phweyland looks like this was introduced by c9b7ea1ed. We can't start writing to sidecar files until the `_image_read_duplicates` function call has completed because of how I resolved this issue the first time round.

Basically, when you load an image that doesn't yet exist in the database, `_image_import_internal` first creates a dummy image record with version 0. When `_image_read_duplicates` is called, if it finds that xmp files are present but that none of them are v0, it updates the version number of that dummy record with the first xmp version it finds.

If you write the xmp file for that dummy record before `_image_read_duplicates` you will always force a v0 xmp to be written even if v0 has been deleted.

I'm not sure how best to resolve this. Any ideas?

**Comment by phweyland** [comment]
Your demonstration is cristal clear.
Thinking aloud ...
The guilty commit intent was to make sure to save the lightroom metadata.
The xmp logic is complex but I think we could ignore lightroom xmp if any normal xmp is detected.
Or the other way, we could call write_xmp only if dt_ligthroom_import() has found a ligthroom xmp... (currently it does gives that info...)

**Comment by elstoc** [comment]
> EDIT: or call write_xmp inside dt_ligthroom_import()...

No because that would still be writing an xmp before we've read all the existing ones.

The place to do it might be within `_image_read_duplicates`, after the loop over `file_iter`, if `count_xmps_processed = 0`. Because that means that no xmp files were read.

**Comment by phweyland** [comment]
> > EDIT: or call write_xmp inside dt_ligthroom_import()...
> 
> No because that would still be writing an xmp before we've read all the existing ones.

I'm afraid not to agree here, because lightroom xmp is considered on the first importation only (then there is no other existing xmp, ... normally).

**Comment by phweyland** [comment]
>  because lightroom xmp is considered on the first importation only 

Reading the code, this is right for dev (applied only once) but not for the image (re-)creation... I think this not correct. If the image already exists in db or if there is at least one xmp, the lightroom xmp has no reason to be treated again.

> The place to do it might be within `_image_read_duplicates`, after the loop over `file_iter`, if `count_xmps_processed = 0`. Because that means that no non-lightroom xmp files were read.

Probably not enough because the lightroom information are not yet available, are they ?

**Comment by elstoc** [comment]
> If the image already exists in db or if there is at least one xmp, the lightroom xmp has no reason to be treated again

Yes I agree but if the image is not already in the db we don't know if there are any xmps until after (or during) `_image_read_duplicates` (though I admit I haven't read the code that reads the lightroom xmp).

> Probably not enough because the lightroom information are not yet available, are they ?

I'm not sure what you mean here but currently you're reading the lightroom xmp _before_ `_image_read_duplicates` is called so if it was available then, it's certainly available _during_ `_image_read_duplicates`.

**Comment by phweyland** [comment]
> you're reading the lightroom xmp _before_ `_image_read_duplicates` is called so if it was available then, it's certainly available _during_ `_image_read_duplicates`.

My bad, you are right. Just consider that `_image_read_duplicates` is also called for existing images. In that case the write_xmp for lightroom data should be avoided.

**Comment by elstoc** [comment]
Agreed

**Comment by elstoc** [comment]
I'm not sure we reached a conclusion here?

**Comment by phweyland** [comment]
> I'm not sure we reached a conclusion here? Or worked out who was going to do the fix.

Agreed.
Difficult to fix anything without creating another issue (a lot of use cases without enough explicit statements).
If you think you have a good fix please go on.

**Comment by elstoc** [comment]
> If you think you have a good fix please go on.

I'm not sure I do. The only simple thing I can think is perhaps to just move the writing of the xmp to after duplicates are read.

**Comment by phweyland** [comment]
> > If you think you have a good fix please go on.
> 
> I'm not sure I do. The only simple thing I can think is perhaps to just move the writing of the xmp to after duplicates are read.

I think that in a normal flow, importing which are not in dt, based on existing xmp, you are right, except that lightroom xmp should be dis-considered if there are duplicates.

Suggestion in line with yours.
Make _image_read_duplicates return the number of xmp found.
If found we can avoid both to call lightroom and write xmp to save lightroom data.
And for the case where the image already exists we don't disturb anything.

How does this sound ?

**Comment by elstoc** [comment]
Sounds reasonable to me.

**Comment by phweyland** [comment]
I make a try then

**Comment by elstoc** [comment]
cool thanks
